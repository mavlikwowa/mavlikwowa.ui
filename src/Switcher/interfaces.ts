export interface SwitcherProps extends React.HTMLProps<HTMLSpanElement> {
  hasChildren?: React.ReactNode;
  checked?: boolean;
}

export interface StyledSwitcherProps {
  hasChildren: boolean;
}
