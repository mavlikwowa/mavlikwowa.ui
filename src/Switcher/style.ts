/* Libraries */
import styled from '@emotion/styled';

/* Intefaces */
import { StyledSwitcherProps } from './interfaces';

export const StyledSwitcher = styled.label<StyledSwitcherProps>`
  ${({ hasChildren }: StyledSwitcherProps) => {
    return `
      display: inline-flex;
      cursor: pointer;
      width: 50px;
      height: 15px;
      background: grey;
      border-radius: 15px;
      align-items: center;
      outline: none;
      
      input {
        display: none;
        
        &:checked ~ span {
         background: ${hasChildren ? 'none' : 'white'};
         transform: translateX(70%) rotate(180deg);
         transition: transform 1s, background 1s, rotate 1s;
         border: ${hasChildren ? 'none' : '1px solid darkgrey'};
        }
        
        &:active ~ span {
          transform: scale(0.8);
        }
      }
      
      span {
        width: 30px;
        height: 30px;
        border-radius: 50%;
        background: ${hasChildren ? 'none' : 'black'};
        transform: translateX(0);
        transition: transform 1s, background 1s;
        border: ${hasChildren ? 'none' : '1px solid black'};
      }
    `;
  }}
`;
