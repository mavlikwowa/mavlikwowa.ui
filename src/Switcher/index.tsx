/* Libraries */
import React from 'react';

/* Interfaces */
import { SwitcherProps } from './interfaces';

/* Styles */
import { StyledSwitcher } from './style';

const Switcher: React.FC<SwitcherProps> = ({ children }: SwitcherProps) => {
  return (
    <StyledSwitcher htmlFor="switcher">
      <input id="switcher" type="checkbox" />
      <span>{children || ''}</span>
    </StyledSwitcher>
  );
};

export default Switcher;
