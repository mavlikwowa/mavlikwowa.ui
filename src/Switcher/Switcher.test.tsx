/* Libraries */
import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';

/* Components */
import Switcher from './index';

/** Setups */
const setUp = () => {
  /* Component */
  const wrapper = shallow(<Switcher />);
  return { wrapper };
};

/* Tests */
describe('Test component Switcher', () => {
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    ({ wrapper } = setUp());
  });

  afterEach(() => {
    if (!wrapper) return;
    wrapper.unmount();
  });

  it('should render a component', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('shouldn`t has a child', () => {
    expect(wrapper.prop('hasChildren')).toBeFalsy();
  });
});
