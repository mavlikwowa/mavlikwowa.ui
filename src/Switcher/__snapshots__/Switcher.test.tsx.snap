// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test component Switcher should render a component 1`] = `
<Styled(label)
  hasChildren={false}
>
  <input
    type="checkbox"
  />
  <span />
</Styled(label)>
`;
